File: src/data/mock/priceHistory.ts
import { PricePoint } from '@/domain/types';
import { MOCK_SPOT_CENTS } from './cards';

/**
 * Générateur déterministe d'historiques de prix. On simule une marche
 * aléatoire bornée (mouvement brownien géométrique discret) ancrée sur le
 * spot actuel, avec une graine par carte pour des courbes stables entre
 * les rendus (pas de scintillement de la Sparkline).
 */

export type Range = '7D' | '30D' | '1Y';

const RANGE_DAYS: Record<Range, number> = { '7D': 7, '30D': 30, '1Y': 365 };

/** PRNG déterministe (mulberry32) à partir d'une graine entière. */
function mulberry32(seed: number) {
  return function () {
    seed |= 0;
    seed = (seed + 0x6d2b79f5) | 0;
    let t = Math.imul(seed ^ (seed >>> 15), 1 | seed);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function seedFromId(id: string): number {
  let h = 2166136261;
  for (let i = 0; i < id.length; i++) {
    h ^= id.charCodeAt(i);
    h = Math.imul(h, 16777619);
  }
  return h >>> 0;
}

/**
 * Reconstruit une série rétrograde : on part du spot actuel (dernier point)
 * et on remonte le temps en appliquant l'inverse des variations, de sorte
 * que le dernier point colle toujours au spot affiché ailleurs dans l'app.
 */
export function generateHistory(cardId: string, range: Range): PricePoint[] {
  const days = RANGE_DAYS[range];
  const spot = MOCK_SPOT_CENTS[cardId] ?? 10000;
  const rand = mulberry32(seedFromId(cardId) ^ days);

  // Volatilité quotidienne plus forte sur les courtes périodes affichées.
  const vol = range === '1Y' ? 0.012 : 0.02;
  const drift = -0.0008; // léger biais haussier en remontant le temps

  const points: PricePoint[] = [];
  const now = Date.now();
  const dayMs = 86400000;
  const step = Math.max(1, Math.round(days / 60)); // ~60 points max

  let value = spot;
  for (let d = 0; d <= days; d += step) {
    const t = now - d * dayMs;
    points.push({ t, cents: Math.round(value) });
    // Remontée du temps : on défait un mouvement brownien géométrique.
    const shock = (rand() - 0.5) * 2 * vol + drift;
    value = value / (1 + shock);
  }

  return points.reverse();
}

/** Pré-calcule les historiques 30J de toutes les cartes (pour l'agrégat). */
export function buildHistoryMap(
  cardIds: string[],
  range: Range,
): Record<string, PricePoint[]> {
  return Object.fromEntries(
    cardIds.map((id) => [id, generateHistory(id, range)]),
  );
}

File: flutter_app/lib/data/price_history.dart
import 'dart:math';
import 'mock_data.dart';

enum ChartRange { d7, d30, y1 }

extension ChartRangeX on ChartRange {
  String get label => switch (this) {
        ChartRange.d7 => '7J',
        ChartRange.d30 => '30J',
        ChartRange.y1 => '1A',
      };
  String get long => switch (this) {
        ChartRange.d7 => '7 jours',
        ChartRange.d30 => '30 jours',
        ChartRange.y1 => '1 an',
      };
  int get days => switch (this) {
        ChartRange.d7 => 7,
        ChartRange.d30 => 30,
        ChartRange.y1 => 365,
      };
}

/// Générateur déterministe d'historiques de prix (marche aléatoire bornée,
/// graine par carte) : courbes stables entre les rebuilds, dernier point
/// toujours ancré au spot affiché ailleurs dans l'app.
abstract final class PriceHistory {
  static int _seed(String id, int salt) {
    var h = 2166136261;
    for (final c in id.codeUnits) {
      h ^= c;
      h = (h * 16777619) & 0xFFFFFFFF;
    }
    return h ^ salt;
  }

  static List<int> series(String cardId, ChartRange range) {
    final spot = MockData.spotCents[cardId] ?? 10000;
    final rand = Random(_seed(cardId, range.days));
    final vol = range == ChartRange.y1 ? 0.012 : 0.02;
    const drift = -0.0008;

    final days = range.days;
    final step = max(1, (days / 60).round());
    final points = <int>[];
    var value = spot.toDouble();

    for (var d = 0; d <= days; d += step) {
      points.add(value.round());
      final shock = (rand.nextDouble() - 0.5) * 2 * vol + drift;
      value = value / (1 + shock);
    }
    return points.reversed.toList();
  }

  /// Série agrégée du portefeuille : somme des historiques des cartes possédées.
  static List<int> portfolioSeries(List<String> cardIds, ChartRange range) {
    List<int>? acc;
    for (final id in cardIds) {
      final s = series(id, range);
      if (acc == null) {
        acc = List<int>.from(s);
      } else {
        for (var i = 0; i < acc.length && i < s.length; i++) {
          acc[i] += s[i];
        }
      }
    }
    return acc ?? const [];
  }
}
